src/audio_out.rs

```diff
@@ -0,0 +1,70 @@
+//! CS43L22 DAC with headphone amp.
+
+use crate::hal::prelude::*;
+
+use crate::hal::gpio::gpioa::{self, PA4};
+use crate::hal::gpio::gpiob::{self, PB6, PB9};
+use crate::hal::gpio::gpioc::{self, PC10, PC12};
+use crate::hal::gpio::gpiod::{self, PD4};
+use crate::hal::gpio::{Alternate, AlternateOD, Output, PushPull, AF4, AF6};
+use crate::hal::i2c::*;
+use crate::hal::i2s::*;
+use crate::hal::rcc::Clocks;
+use crate::hal::stm32::{I2C1, SPI3};
+
+use embedded_hal::blocking::delay::DelayMs;
+
+pub use cs43l22;
+
+pub struct AudioOut {
+    pub cs43l22: cs43l22::CS43L22<
+        I2c<I2C1, (PB6<AlternateOD<AF4>>, PB9<AlternateOD<AF4>>)>,
+        PD4<Output<PushPull>>,
+    >,
+    pub i2s: I2s<
+        SPI3,
+        (
+            PC10<Alternate<AF6>>,
+            PA4<Alternate<AF6>>,
+            PC12<Alternate<AF6>>,
+            NoSdExt,
+        ),
+    >,
+}
+
+impl AudioOut {
+    pub fn new<DELAY: DelayMs<u8>>(
```

Comment:
It would make more sense to pass the specific pins rather than the full gpio blocks. As it is all blocks get moved into the new method and then you can't any pins for anything else.

Reply:
It’s done like this as that’s how it was done for the LED and accelerometer setup in this crate. I believe passing parts only moves the used pins after the function returns?